Finances/load.js

```diff
@@ -46,34 +89,89 @@ CampaignFinanceLoader.prototype.loadCampaignFinances = async function(finance_ar
-
-    const candidateSearch = await this.CandidateSearch.findOne({
-      where
-    })
-    if(candidateSearch == null){
-      console.log(`Could not find ${JSON.stringify(where, null, 4)}`)
-    }
-    const contributions = finance_object.contributions ? finance_object.contributions : null;
-    const expenditures = finance_object.expenditures ? finance_object.expenditures : null;
-
-    if(candidateSearch){
-      const campaign_finance = await this.CampaignFinance.findOrCreate({
+    };
+
+    const state = await this.State.findOne({
+      where: {
+        name: state_name
+      }
+    });
+    if(!state) throw new Error('invalid state: '+state_name);
+    const chamber = await this.Chamber.findOne({
+      where: {
+        state_id: state.state_id,
+        level
+      }
+    });
+    if(!chamber) throw new Error('invalid chamber: '+level);
+    const district = await this.District.findOne({
+      where: {
+        chamber_id: chamber.chamber_id,
+        number: where.district
+      }
+    });
+    if(!district) throw new Error('invalid district');
+    const office = await this.Office.findOne({
+      where: {
+        district_id: district.district_id
+      }
+    });
+    if(!office) throw new Error('invalid office');
+    const election = await this.Election.findOne({
+      where: {
+        office_id: office.office_id,
+        year: election_year
+      }
+    });
+
+    // sequelize.where(sequelize.fn('LOWER', sequelize.col('firstname')), 'LIKE', '%' + search.toLowerCase() + '%')
+
+    const candidates = await this.Candidate.findAll({
+      where: {
+        first_name: sequelize.where(sequelize.fn('LOWER', sequelize.col('first_name')), Op.eq, first_name.toLowerCase()),
+        last_name: sequelize.where(sequelize.fn('LOWER', sequelize.col('last_name')), Op.eq, last_name.toLowerCase())
+      }
+    });
+    if(!candidates.length) {
+      console.log('no candidates found for: ',parsedName, candidates);
+      noMatch.push({...finance_object, first_name, last_name});
+    } else {
+      const candidacies = await this.Candidacy.findAll({
         where: {
-          candidacy_id: candidateSearch.candidacy_id,
-        },
-        defaults:{
-
-          contributions: contributions,
-          expenditures:expenditures
+          election_id: election.election_id,
+          candidate_id: {
+            [Op.in]: candidates.map(x=>x.candidate_id)
+          }
         }
-      })
+      });
+
+      let [candidacy] = [...candidacies].sort((a,b) => {
+        if(a.election_type == 'general') return -1;
+        if(b.election_type == 'general') return 1;
+        
+        return -1;
+      });
+
+      if(candidacy) {
+        console.log('found it!', candidacy);
+        const contributions = finance_object.contributions ?? null;
+        const expenditures = finance_object.expenditures ?? null;
+
+        await this.CampaignFinance.findOrCreate({
+          where: {
+            candidacy_id: candidacy.candidacy_id,
```

Comment:
Just wondering if this handles duplicate finance records OK? For some reason Arizona, for example, will return multiple return multiple responses on the same candidate, with identical information. 

On my first crack at it findOrCreate threw a fit over attempting to create multiple records with the same primary key. 

Making a note (for myself) to test this on AZ to make sure dupes don't break anything. 

Maybe de-duping should be done in the getters for campaign finance? Not sure which is ideal.

Reply:
I've just been calling the campaign finance scrapers "getters." Pretty silly but seemed apt.